polar-core/src/partial/simplify.rs

```diff
@@ -1,217 +1,128 @@
-                        }));
+                match operations.get(0).unwrap() {
+                    Operation {
+                        operator: Operator::And,
+                        args,
+                    } if args.len() == 1 => fold_term(args.get(0).unwrap().clone(), self),
+
+                    Operation {
+                        operator: Operator::Unify,
+                        args,
+                    } if args.iter().any(is_this_arg) => {
+                        let mut args = args
+                            .iter()
+                            .filter(|arg| !is_this_arg(arg))
+                            .cloned()
+                            .collect::<TermList>();
+                        assert_eq!(args.len(), 1);
+                        args.pop().unwrap()
                     }
-                    (false, true) => {
-                        args.push(term!(Operation {
-                            operator: operation.operator,
-                            args: vec![term!(left.clone()), term!(dot_op.clone())]
-                        }));
-                    }
-                    (_, _) => panic!("invalid"),
+                    _ => fold_term(t, self),
                 }
-            } else {
-                args.push(term!(operation.clone()))
             }
+            _ => fold_term(t, self),
         }
-
-        operations.push(Operation {
-            operator: Operator::And,
-            args,
-        });
-    }
-}
-
-fn not_this_arg(operation: &Operation) -> Option<Term> {
-    let left = operation.args.get(0).unwrap();
-    let right = operation.args.get(1).unwrap();
-
-    match (is_this_arg(left.value()), is_this_arg(right.value())) {
-        (false, true) => Some(left.clone()),
-        (true, false) => Some(right.clone()),
-        _ => None,
     }
-}
-
-fn is_this_arg(value: &Value) -> bool {
-    matches!(value, Value::Variable(sym) if sym.0 == "_this")
-}
-
-// partial(_x_5) { partial(_value_1_6) { _this > 0, _this > 1 } = _this.a }
 
-// Take partial(_this = ?) and output ?.
-fn simplify_unify_partials(term: Term, _: &Bindings) -> Term {
-    if let Value::Partial(p) = term.value() {
-        let operator = p.operations().first().unwrap().operator;
-        let is_unify = matches!(operator, Operator::Unify);
-
-        if p.operations().len() == 1 && is_unify {
-            let op = p.operations().first().unwrap();
-
-            match not_this_arg(op) {
-                Some(term) => term,
-                None => term.clone(),
+    fn fold_operation(&mut self, o: Operation) -> Operation {
+        /// Given `_this` and `x`, return `x`.
+        /// Given `_this.x` and `_this.y`, return `_this.x.y`.
+        fn sub_this(term: &Term, replacement: &Term) -> Term {
+            match (term.value(), replacement.value()) {
+                (Value::Variable(v), _) if v.is_this_var() => replacement.clone(),
+                (
```

Comment:
I believe that it is possible, but we couldn't easily construct a test case. But the logic of this function is that of a filter, not a checker; we're only interested in the cases where we do have an arg that is `_this`, so it would seem wrong to assert that.